dali/python/nvidia/dali/external_source.py

```diff
@@ -340,31 +344,42 @@ class ExternalSource():
     callback in parallel. You can specify the number of workers by passing ``py_num_workers``
     into pipeline's constructor.
 
-    When ``parallel`` is set to True, ``source`` must return NumPy/MXNet/PyTorch CPU array,
-    TensorCPU, or tuple/list of these types with length matching num_outputs.
+    When ``parallel`` is set to True, samples returned by ``source`` must be
+    NumPy/MXNet/PyTorch CPU arrays or TensorCPU instances.
 
-    Only callables that accept one argument (:meth:`~nvidia.dali.types.SampleInfo` objects that
-    represent the index of the requested sample) can be used as ``source`` when ``parallel`` is
-    set to True. It can be a function or an object implementing ``__call__`` operator, which
-    allows to add an initial state to the object instance.
+    Acceptable sources depend on the value specified for ``batch`` parameter.
 
-    Keep in mind, that **copies** of the ``source`` will be distributed between Python workers,
-    and no global state can be shared between them.
+    If batch is set to True, source must be a callable (a function or an object with __call__ method)
+    that accepts exactly one argument (:meth:`~nvidia.dali.types.SampleInfo` objects that
+    represent the index of the requested sample).
+    If ``batch=True``, then ``source`` can be either a callable, an iterable or a generator function.
+    Callable in batch mode must accept exactly one argument - an integer that represents the index of the
+    batch within the epoch that the callable should return.
 
-    The ``source`` callback must raise StopIteration when the end of data is reached.
+    Irrespective of ``batch`` value, callables should be statless and be able to produce
+    requested sample or batch solely based on the SampleInfo instance or index in batch, so that they can
+    be run in parallel in a number of workers. When batch is set to True, callables performance might
+    especially benefit from increasing ``prefetch_queue_depth`` so that a few next batches can be
+    computed in parallel.
 
-    Setting ``parallel`` to True makes the external source work in per-sample mode.
-    If ``batch`` was not set it is set to False.
+    Iterator or generator will get assigned a single worker that will iterate over it.
+
+    The ``source`` callback must raise StopIteration when the end of data is reached.
 
 `prefetch_queue_depth` : int, option, default = 1
     When run in ``parallel=True`` mode, specifies the number of batches to be computed in advance and stored
     in the internal buffer, otherwise parameter is ignored.
+
+`bytes_per_minibatch_hint` : int, default = None
+    Hint for the initial size of each buffer accommodating (part of) the batch returned from worker
+    in parallel mode. You may want to specify custom value to lower the number of resizes necessary when
```

Comment:
Is there any reasonable way to learn what is the right value? Or just some low level debugger is needed?